### ISSUE TITLE
fine-tune issue and audio not generate

### ISSUE THREAD

**Issue by saifulislam79** [ekwek1/soprano-factory#7]
Thank you for your great work. I have generate speech token using your encode but for inference audio encode i need to modify codec code-base before inference audio codec.

modify script before inference audio code inferece audio process using below function
`https://github.com/ekwek1/soprano-factory/blob/main/generate_dataset.py`

```
def _to_float32_waveform(self, audio_np: np.ndarray) -> torch.Tensor:
        """
        #Convert numpy audio (int16/int32/uint8/float) to torch.float32 in [-1, 1],
        #and return shape [1, T] (mono) or [C, T] (multi-channel).
        """
        # If scipy returns shape [T, C], convert to [C, T]
        if audio_np.ndim == 2:
            audio_np = audio_np.T

        if audio_np.dtype == np.int16:
            audio = torch.from_numpy(audio_np).to(torch.float32) / 32768.0
        elif audio_np.dtype == np.int32:
            audio = torch.from_numpy(audio_np).to(torch.float32) / 2147483648.0
        elif audio_np.dtype == np.uint8:
            # uint8 PCM is usually 0..255 with 128 as zero
            audio = (torch.from_numpy(audio_np).to(torch.float32) - 128.0) / 128.0
        else:
            # already float (float32/float64) - assume it's in [-1, 1] or close
            audio = torch.from_numpy(audio_np).to(torch.float32)

        # Ensure shape [C, T]
        if audio.ndim == 1:
            audio = audio.unsqueeze(0)

        return audio
```

I have training `10000` steps but loss increase,
```
validation text loss: 14.8595   validation audio loss: 4.2837   validation acc: 0.1632
text loss: 15.175 | audio loss: 4.214 | acc: 0.1664 | lr: 5.00e-04 | norm: 0.602 | time: 284.53 ms | 230330.01 t/s:  90%|▉| 9000/10000 [42:04<0validation text loss: 15.1198   validation audio loss: 4.2776   validation acc: 0.1631
text loss: 14.151 | audio loss: 4.126 | acc: 0.1810 | lr: 3.88e-04 | norm: 0.621 | time: 271.59 ms | 241307.24 t/s:  92%|▉| 9250/10000 [43:15<0validation text loss: 14.1243   validation audio loss: 4.2479   validation acc: 0.1656
text loss: 14.148 | audio loss: 4.136 | acc: 0.1766 | lr: 2.75e-04 | norm: 0.475 | time: 266.49 ms | 245926.02 t/s:  95%|▉| 9500/10000 [44:27<0validation text loss: 14.1074   validation audio loss: 4.2051   validation acc: 0.1705
text loss: 13.899 | audio loss: 4.169 | acc: 0.1757 | lr: 1.63e-04 | norm: 0.488 | time: 276.66 ms | 236886.52 t/s:  98%|▉| 9750/10000 [45:40<0validation text loss: 13.8969   validation audio loss: 4.1808   validation acc: 0.1724
text loss: 13.975 | audio loss: 4.188 | acc: 0.1833 | lr: 5.09e-05 | norm: 0.428 | time: 288.29 ms | 227322.87 t/s: 100%|▉| 9999/10000 [46:51<0validation text loss: 13.9734   validation audio loss: 4.1624   validation acc: 0.1755
text loss: 13.975 | audio loss: 4.061 | acc: 0.1874 | lr: 5.04e-05 | norm: 0.461 | time: 438.44 ms | 149475.87 t/s: 100%|█| 10000/10000 [46:51<
```

But audio generate garbage by the by i am working on bangla language.

[out.wav](https://github.com/user-attachments/files/24744310/out.wav)

if you have any suggestion please let me know i am waiting for your response.

**Comment by yukiarimo** [comment]
Codec is fine, and is trainable. LM training is currently broken

**Comment by saifulislam79** [comment]
@yukiarimo how may steps you have trained 1000 hours data and is it possible share training logs

**Comment by mhenrichsen** [comment]
> Codec is fine, and is trainable. LM training is currently broken

@yukiarimo do you have any code to training the encoder?

**Comment by Sedherthe** [comment]
@yukiarimo How did you train the codec for this?

**Comment by yukiarimo** [comment]
I wrote a training code for it. Will share it later

**Comment by Sedherthe** [comment]
@yukiarimo the codec that you trained, can you share more details on wha encoder-decoder arch you used? How are you verifying it trained well?

**Comment by lukedupin** [comment]
interested to get some training help. I went through creating a data sent and just got non-sense out.

**Comment by yukiarimo** [comment]
Until the training loop for LLM isn’t fixed, there’s no point in training anything :(

**Comment by mhenrichsen** [comment]
@yukiarimo I have working training code for the LLM. Mind sharing the code for encoder/decoder? https://gist.github.com/mhenrichsen/1809396c04ca43b14c9a0cdcc7926d6d 

The code is training for voice cloning, but should be easily adapted.

**Comment by Nighthawk42** [comment]
Any updates on this @yukiarimo and @mhenrichsen - I've been experimenting with mhenrichsen's gist and have made some progress. I'm now getting my data's voice as gibberish instead of the default models. It's still missing something...
